File: fs/BeginnersSelection/04_Coins.fs
// https://atcoder.jp/contests/abs/tasks/abc087_b

// (stdin.ReadLine(), stdin.ReadLine(), stdin.ReadLine(), stdin.ReadLine())
// |> fun (a, b, c, x) -> (a |> int, b|> int, c|> int, x|> int)
// |> fun (a, b, c, x) -> 
//   [0..(min a (x / 500))]
//   |> List.map (fun i ->
//     [0..(min b ((x - 500 * i) / 100))]
//     |> List.filter (fun j -> x - 500 * i - 100 * j <= 50 * c)
//     |> List.length
//   )
//   |> List.sum
//   |> printf "%d"

// ----------

// (stdin.ReadLine(), stdin.ReadLine(), stdin.ReadLine(), stdin.ReadLine()) |> fun (a, b, c, x) -> (a |> int, b|> int, c|> int, x|> int) |> fun (a, b, c, x) ->  [0..(min a (x / 500))] |> List.map (fun i -> [0..(min b ((x - 500 * i) / 100))] |> List.filter (fun j -> x - 500 * i - 100 * j <= 50 * c) |> List.length) |> List.sum |> printf "%d"

// https://atcoder.jp/contests/abs/submissions/9416476

// ----------

(stdin.ReadLine(), stdin.ReadLine(), stdin.ReadLine(), stdin.ReadLine())
|> fun (a, b, c, x) -> (a |> int, b|> int, c|> int, x|> int)
|> fun (a, b, c, x) -> 
  [0..(min a (x / 500))]
  |> List.sumBy (fun i ->
    [0..(min b ((x - 500 * i) / 100))]
    |> List.filter (fun j -> x - 500 * i - 100 * j <= 50 * c)
    |> List.length
  )
  |> printf "%d"

File: ex/BeginnersSelection/04_Coins.ex
# https://atcoder.jp/contests/abs/tasks/abc087_b

defmodule Main, do:
  def main(), do:
    [IO.read(:line), IO.read(:line), IO.read(:line), IO.read(:line)]
    |> Enum.map(&String.trim/1)
    |> Enum.map(&String.to_integer/1)
    |> (fn t -> apply(fn a, b, c, x ->
      (0..Enum.min([a, floor(x / 500)]))
      |> Enum.map(fn i ->
        (0..Enum.min([b, floor((x - 500 * i) / 100)]))
        |> Enum.filter(fn j -> x - 500 * i - 100 * j <= 50 * c end)
        |> Enum.count
      end)
      |> Enum.sum
    end, t) end).()
    |> IO.puts

# https://atcoder.jp/contests/abs/submissions/21757220
